### ISSUE TITLE
FGSM & PGD nit picks

### ISSUE THREAD

**Issue by zroe1** [zroe1/xlab-ai-security#1]
1. Use `epsilon` instead of `e` as your budget variable in `FGSM_generator`
2. This is my bad for not flagging this earlier but get rid of the `def process_image(path)` and just have the CNN predict based on the raw image. While preprocessing in this way does increase accuracy, it makes it much more difficult to run automated attacks. For example AutoAttack by default will assume that the image is between 0 and 1. Likewise, CW assumes the image is between 0 and 1. It also makes the budgets less meaningful. If you set $\epsilon=8/255$ for an image that is between 0 and 1, that is very different than setting an setting the same epsilon for an image that is normalized in the the way that you have done it.
 
Currently the min and max of your image suggest a much larger range:
```python
torch.min(process_image('frog.jpg')), torch.max(process_image('frog.jpg'))
```
```markdown
(tensor(-0.9608), tensor(0.9922))
```

This will require you to retrain your CNN so sorry about not flagging this earlier :( 

3. Use LaTeX inside of images of equations. Inside markdown blocks of a Jupyter notebook this is supported

E.g.,
```markdown
$$
y = m \cdot x + b
$$
```

Will render as
$y = m \cdot x + b$

(but it will be centered within the notebook--not here for some reason)

4. For `FGSM_generator` give the equation for loss that you want people to implement (there a number of reasonable choices that one could make for this). I notice that you load `loss_fn = torch.nn.CrossEntropyLoss()` earlier but I missed this when I was filling in the notebook. I used 

I was able to convince myself that your code for `FGSM_generator` is correct so this shouldn't require further modification.

**Comment by zroe1** [comment]
In your comment for building the model you have
```
#Conv1, pooling, dropout, conv2, pooling, flatten, linear1, dropout, linear2
```
But based on your solution I think you mean to have
```
#Conv1, pooling, dropout, conv2, pooling, flatten, linear1, relu, dropout, linear2
```
(with the relu after linear 1)

**Comment by zroe1** [comment]
I've written my own code and looked at yours and I've convinced myself that your code for `IGSM_generator` is working correctly

**Comment by zroe1** [comment]
For section `# 5. Making an adversarial image using PGD` I would give a text description of what is different between IGSM and PGD so it is clear what the user will have to implement.

**Comment by zroe1** [comment]
Also checked the pgd and wrote my own code and your solution lgtm

I really like how you factored this into it's own function

```python
def add_noise(img, stdev=0.01, mean=0):
    noise = torch.randn_like(img) * stdev + mean
    return img + noise
```

I think this is really helpful because it abstracts away something that isn't as important.

Additionally I think it I would be useful to do the same thing for a clamping function. E.g.,

```python
def clamp_img(... arguments ...):
     ### your code here ###
```

The clamping part is one of the hardest parts to get right and I think it makes sense to make this it's own step, where students fill in the code and then run tests before moving onto IGSM_generator and PGD

**Comment by zroe1** [comment]
One last tiny bug:

The return line for the last function should be
```
return x ** (1/p)
```
rather than 


```
return x ** 1/p
```
The above just does $\frac{(x^1)}{p}$ rather than $x^{\frac{1}{p}}$

**Comment by zroe1** [comment]
I'm making some little wording changes here:


old:
```
To save having to store the model architecture as a file, only the weights can be stored, and the model architecture can be replicated as follows.
```

new:
```
We have trained the model for you and saved the weights in `CNN_weights.pth`. In order to load the weights, you will have to fill in the architecture below. 
```

At some point, we will have to give people instructions for how to load the weights but I am not worried about that for now.

**Comment by zroe1** [comment]
Not sure what you mean by this at the top of the notebook? I'm guessing you are talking about fgsm & pgd but this reads weirdly to me.

```
Our utility library contains the majority of the following functions. They will not have to be implemented in future notebooks.
```

**Comment by zroe1** [comment]
Lol this is super nit picky but I wonder if the comments would look cleaner if instead of something like this

```python
###TODO
#Write a PyTorch image composition to resize and convert to a tensor
transform = None;
###
```

it looked something like this

```python
######### YOUR CODE STARTS HERE ######### 
# Write a PyTorch image composition to resize and convert to a tensor
transform = None;
########## YOUR CODE ENDS HERE ########## 
```

If you disagree, you can leave it how it is--it honestly doesn't really matter.

**Comment by zroe1** [comment]
In the hint for task two I'm guess you want to say 

```python
Compose([Transform1, Transform2])
```

rather than 

```python
Compose(Transform1, Transform2)
```

I fixed this for you but make sure that this seems right

**Comment by zroe1** [comment]
@pgarodia1 I made a bunch of changes and put them in a new file below for you to review and make sure you are okay with everything:

https://github.com/zroe1/xlab-ai-security/blob/main/working/FGSM_IGSM_PGD.ipynb
